internal-tooling/performPerformanceTest.ts

```diff
@@ -0,0 +1,189 @@
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import yargs from 'yargs';
+import * as uuid from 'uuid';
+import {execSync} from 'child_process';
+import {unlinkSync} from 'fs';
+import {Storage} from '../src';
+import {performance} from 'perf_hooks';
+import {parentPort} from 'worker_threads';
+import path = require('path');
+
+const TEST_NAME_STRING = 'nodejs-perf-metrics';
+const DEFAULT_NUMBER_OF_WRITES = 1;
+const DEFAULT_NUMBER_OF_READS = 3;
+const DEFAULT_BUCKET_NAME = 'nodejs-perf-metrics';
+const DEFAULT_SMALL_FILE_SIZE_BYTES = 5120;
+const DEFAULT_LARGE_FILE_SIZE_BYTES = 2.147e9;
+const BLOCK_SIZE_IN_BYTES = 1024;
+const NODE_DEFAULT_HIGHWATER_MARK_BYTES = 16384;
+
+export interface TestResult {
+  op: string;
+  objectSize: number;
+  appBufferSize: number;
+  libBufferSize: number;
+  crc32Enabled: boolean;
+  md5Enabled: boolean;
+  apiName: 'JSON' | 'XML';
+  elapsedTimeUs: number;
+  cpuTimeUs: number;
+  status: '[OK]';
+}
+
+const randomInteger = (minInclusive: number, maxInclusive: number) => {
+  // Utilizing Math.random will generate uniformly distributed random numbers.
+  return (
+    Math.floor(Math.random() * (maxInclusive - minInclusive + 1)) + minInclusive
+  );
+};
+
+const argv = yargs(process.argv.slice(2))
+  .options({
+    bucket: {type: 'string', default: DEFAULT_BUCKET_NAME},
+    small: {type: 'number', default: DEFAULT_SMALL_FILE_SIZE_BYTES},
+    large: {type: 'number', default: DEFAULT_LARGE_FILE_SIZE_BYTES},
+    projectid: {type: 'string'},
+  })
+  .parseSync();
+
+async function main() {
+  const results = await performWriteReadTest();
+  parentPort?.postMessage(results);
+}
+
+async function performWriteReadTest(): Promise<TestResult[]> {
+  const results: TestResult[] = [];
+  const fileName = generateRandomFileName();
+  const sizeInBytes = generateRandomFile(fileName);
+  const checkType = randomInteger(0, 2);
+
+  const stg = new Storage({
+    projectId: argv.projectid,
+  });
+
+  let bucket = stg.bucket(argv.bucket);
+  if (!(await bucket.exists())[0]) {
+    await bucket.create();
+  }
+
+  for (let j = 0; j < DEFAULT_NUMBER_OF_WRITES; j++) {
+    let start = 0;
+    let end = 0;
+
+    const iterationResult: TestResult = {
+      op: 'WRITE',
+      objectSize: sizeInBytes,
+      appBufferSize: BLOCK_SIZE_IN_BYTES,
+      libBufferSize: NODE_DEFAULT_HIGHWATER_MARK_BYTES,
+      crc32Enabled: false,
+      md5Enabled: false,
+      apiName: 'JSON',
+      elapsedTimeUs: 0,
+      cpuTimeUs: -1,
+      status: '[OK]',
+    };
+
+    bucket = stg.bucket(argv.bucket, {
+      preconditionOpts: {
+        ifGenerationMatch: 0,
```

Comment:
I believe this should go on bucket.upload() call no?

Reply:
When we implemented preconditions due to constraints with common (i.e. we couldn't pass preconditions per function), we had to pass them at the constructor level. One of the upcoming refactors with internalizing common is adding the ability to pass these per function. If you look at the code below these lines you will see I re-declare the bucket as necessary to remove the preconditions where appropriate.

Reply:
I literally just spoke with @shaffeeullah and she clarified this as well (in-person no way!!); thanks for clarifying, it will get better just didn't realize this was a trade off with not internalizing common. It's okay, not great but at least it's possible to pass in the precondition.